File: 0501~1000/0844. Backspace String Compare/main.py
"""
Source: https://leetcode.com/problems/backspace-string-compare/
Date: 2023/1/23
Skill:
Runtime: 29 ms, faster than 90.42%
Memory Usage: 13.8 MB, less than 66.28%
Time complexity:
Space complexity:
Constraints:
    1 <= s.length, t.length <= 200
"""

import math
from typing import List
from collections import defaultdict, Counter
from heapq import heapify, heappush, heappop, nsmallest

import heapq
import functools


class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        def build(string):
            res = []
            for c in string:
                if c == "#":
                    if res:
                        res.pop()
                else:
                    res.append(c)

            return res

        return build(s) == build(t)

File: 0501~1000/0844. Backspace String Compare/main.cpp
/**
 * Source: tinyurl.com/yt5trs8t
 * Date: 2023/10/19
 * Skill:
 * Runtime: 0 ms, faster than 100.00% of C++ online submissions
 * Memory Usage: 6.62 MB, less than 39.46% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>

using namespace std;

#define ll long long
typedef pair<int, int> pairs;
typedef unsigned long long ULL;
using PULL = pair<ULL, ULL>;


class Solution {
public:
    bool backspaceCompare(string s, string t) {
        string ss, tt;
        for (auto c: s) {
            if (c == '#') {
                if (!ss.empty())
                    ss.pop_back();
            }
            else
                ss.push_back(c);
        }
        for (auto c: t) {
            if (c == '#') {
                if (!tt.empty())
                    tt.pop_back();
            }
            else
                tt.push_back(c);
        }
        return ss == tt;
    }
};


int main() {
    Solution s;
    vector<int> nums{3,3,2};
    vector<vector<int>> arr{{1,10}, {3,3}};
    function<int(int, int)> test;
    test = [&](int a, int b) {
        if (a > 0)
            return test(a - 1, 0);
        else
            return 0;
    };
}